File: src/Contest157.java
import java.util.*;

class Solution {
    public long sumOfLargestPrimes(String s) {
        Set<Integer> primes = new HashSet<>();
        
        // Generate all substrings and check for primes
        for (int i = 0; i < s.length(); i++) {
            for (int j = i + 1; j <= s.length(); j++) {
                String substring = s.substring(i, j);
                int num = Integer.parseInt(substring); // Automatically handles leading zeros
                if (isPrime(num)) {
                    primes.add(num);
                }
            }
        }
        
        // Convert to list and sort in descending order
        List<Integer> primeList = new ArrayList<>(primes);
        primeList.sort(Collections.reverseOrder());
        
        // Sum the first 3 (or fewer if less than 3 exist)
        long sum = 0;
        for (int i = 0; i < Math.min(3, primeList.size()); i++) {
            sum += primeList.get(i);
        }
        
        return sum;
    }
    
    private boolean isPrime(int n) {
        if (n <= 1) return false;
        if (n <= 3) return true;
        if (n % 2 == 0 || n % 3 == 0) return false;
        
        // Check for factors of the form 6k ± 1
        for (int i = 5; i * i <= n; i += 6) {
            if (n % i == 0 || n % (i + 2) == 0) {
                return false;
            }
        }
        return true;
    }
}

File: biweekly/contest_157.rb
require 'set'

class Contest157
  # @param {String} s
  # @return {Integer}
  def sum_of_largest_primes(s)

    primes = Set.new

    (0...s.length).each do |i|
      (i + 1..s.length).each do |j|
        substring = s[i...j]
        num = substring.to_i
        primes.add(num) if prime?(num)
      end
    end

    primes.to_a.sort.reverse.take(3).sum
  end

  private

  def prime?(n)
    return false if n <= 1
    return true if n <= 3
    return false if n % 2 == 0 || n % 3 == 0

    # Check for divisors of the form 6k ± 1
    i = 5
    while i * i <= n
      return false if n % i == 0 || n % (i + 2) == 0
      i += 6
    end
    true
  end


  def running_sum(nums)
    nums.each_with_index do |num, index|
      nums[index] = nums[index - 1] + num if index > 0
    end
    nums
  end
end
